design.md

```diff
@@ -0,0 +1,75 @@
+# Teleport assessment design document
+
+## Goal
+Implement a prototype job worker service that provides an API to run arbitrary
+Linux processes.
+
+## Architecture
+The system is made up of 3 components. The worker library, the grpc api and the client/cli interface to run commands
+* The worker library provides methods to start, stop and get the status of a running job.
+A method is also provided to stream the output of a job. The api can also be queried for all previous output of a running job. The server also provides a stream for clients to listen to job output.
+* The grpc api is a grpc service that sits on top of the worker library. It provides methods to start, stop and get the status of a running job remotely. 
+* The client/cli is an interface for users to start, stop and get the status of a running job on a remote server.
+
+## Worker library API
+The worker library consists of 3 components. The supervisor, the job worker and the process.
+### Supervisor API
+The supervisor is responsible for managing all jobs in the system. It provides apis to start, stop, get the output and get the status of a job. 
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Add(cmd string, args []string, opts...Options)` - Adds a new job without running it.
+* `Start(id string)` - Starts a previously added job.
+* `Stop(id string)` - Stops a running job by sending the job's process group a `SIGTERM` signal.
+* `Signal(id string, sig syscall.Signal)` - Sends an os signal to a job's process group.
+* `Run()` - Run the supervisor and listen for incoming job requests.
+* `WatchOutput(id string, c chan *pb.JobOutput)` - Add an output to channel to subscribe to process output. All previous output, if any, from the process is replayed before sending new messages to the channel.
```

Comment:
nit: this slightly breaks the boundary between gRPC API and the library
keep the proto types at the gRPC service layer